### ISSUE TITLE
full backward hook won't release grad_outputs

### ISSUE THREAD

**Issue by harryhan618** [pytorch/pytorch#74831]
### 🐛 Describe the bug

In file torch/utils/hooks.py,
If a full_backward_hook is registered, `torch.utils.hooks.BackwardHook` object won't automatically deconstruct, causing its `self.grad_outputs` keeping alive. But usually when no full_backward_hook is registered, those grad_outputs Tensor would disappear inside autograd engine.

This would cause those grad_output memory being kept in the memory. And more full backward hooks are registered on different model submodules, more memory will be kept and don't get released.

Following code proves `torch.utils.hooks.BackwardHook` in torch.nn.modules.Module._call_impl won't get deconstructed.
``` python
import torch
import torch.nn as nn
import gc
from pytorch_memlab import MemReporter

model = nn.Sequential(nn.Linear(5, 4, bias=False), nn.Linear(4,5, bias=False)).cuda()
reporter = MemReporter(model)
model[0].register_full_backward_hook(lambda m, gin, gout: None)
x = torch.rand(7, 5).cuda()
y = model(x).sum()
y.backward()

reporter.report()
print([obj for obj in gc.get_objects() if isinstance(obj, torch.utils.hooks.BackwardHook)])
```

print output is:
```
Element type                                            Size  Used MEM
-------------------------------------------------------------------------------
Storage on cuda:0
0.weight                                              (4, 5)   512.00B
0.weight.grad                                         (4, 5)   512.00B
1.weight                                              (5, 4)   512.00B
1.weight.grad                                         (5, 4)   512.00B
Tensor0                                               (7, 5)   512.00B
Tensor1                                                 (1,)   512.00B
Tensor2                                               (7, 4)   512.00B
-------------------------------------------------------------------------------
Total Tensors: 144      Used Memory: 3.50K
The allocated memory on cuda:0: 3.50K
-------------------------------------------------------------------------------
[<torch.utils.hooks.BackwardHook object at 0x7f52ab3decd0>]
```

### Versions

pytorch 1.10

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

**Comment by albanD** [comment]
What do you mean by "it won't automatically deconstruct"? That object is local to the call of the Module so it is freed properly.
If it is involved in a ref cycle, you might need a `gc.collect()` for it to be detected as dead and removed. Is that what you mean?

**Comment by harryhan618** [comment]
@albanD Hello! Sorry I didn't express myself well. I mean during training, after each iteration, normally those intermediate grad_output would automatically get released, if I don't register any full backward hook.  

But if I use `register_full_backward_hook`, the created `torch.utils.hooks.BackwardHook` during `_call_impl` still keeps alive after I call `loss.backward()` in each iteration. This leads to the fact that `BackwardHook` still holds its `self.grad_outputs`, which means that intermediate grad_output would not automatically get released. 

This means that using full backward hook will cause more memory usage. And extra memory is used to hold intermediate grad_output, which is meaningless.

**Comment by albanD** [comment]
Thanks for the details.
- They do get released. They are just kept alive a bit longer alongside the autograd graph.
- We could drop these earlier (just after calling the user function) but that might lead to some weird behavior when there are multiple backward calls done on a single graph.
- This design will not work well for concurrent backward either but that is not new.

I think as long as we can make sure that the current error message is preserved and that we do get nice error message for the new case (what happens if the hook is called after the grad_outputs are cleared), this is fine.

**Comment by harryhan618** [comment]
@albanD  Sorry, could you check again whether `torch.utils.hooks.BackwardHook` does get released after each 'loss.backward()'? Because in each iteration, after `loss.backward()`, `torch.cuda.memory_allocated()` and `pytorch_memlab.MemReporter` both show extra memory occupation than expected. And I check again that these memory is  `grad_ouput` of BP pass through comparing python `id()` in BP and in `gc.get_objects()`.

So what I mean is that using full backward hook will cause more memory occupation. This is why I file this issue.

**Comment by harryhan618** [comment]
@albanD @mruberry @jbschlosser @walterddr @kshitij12345
Execuse me, is this issue still under review?

**Comment by albanD** [comment]
Hi,

I don't think there is anything new beyond my comment above.
The code sample you share does have the expected behavior: if `y` is still in scope, then it keeps the autograd graph alive and thus the hook.
As mentioned above, we could clean these buffers earlier. We need to make sure we get the right behavior when multiple backward are involved though. If you're interested in contributing this change, I am happy to discuss it more with you / review the PR.

**Comment by soulitzer** [comment]
Fixed by https://github.com/pytorch/pytorch/issues/74831
